include/seqan3/index/fm_index.hpp

```diff
@@ -0,0 +1,336 @@
+ *
+ * Here is an example using a collection of strings (e.g. a genome with multiple chromosomes or a protein database):
+ *
+ * TODO: Coming soon. Stay tuned!
+ *
+ * There is also a history iterator, i.e. an iterator that stores its previous states on a stack such that going down
+ * an edge can be undone. Please take a look at the documentation of seqan3::fm_index_history_iterator::up() since it
+ * does not undo all operations.
+ *
+ * ### Choosing an index implementation
+ *
+ * The underlying implementation of the FM Index (Rank data structure, sampling rates, etc.) can be specified ...
+ */
+template <forward_range_concept text_t, fm_index_traits_concept fm_index_traits = fm_index_default_traits>
+//!\cond
+    requires alphabet_concept<innermost_value_type_t<text_t>>
+        && std::is_same_v<typename underlying_rank<innermost_value_type_t<text_t>>::type, uint8_t>
+//!\endcond
+class fm_index
+{
+protected:
+    //!\privatesection
+
+    //!\brief The type of the underlying SDSL index.
+    using sdsl_index_type = typename fm_index_traits::sdsl_index_type;
+    /*!\brief The type of the reduced alphabet type. (The reduced alphabet might be smaller than the original alphabet
+     *        in case not all possible characters occur in the indexed text.)
+     */
+    using comp_char_type = typename sdsl_index_type::alphabet_type::comp_char_type;
+
+    //!\brief Underlying index from the SDSL.
+    sdsl_index_type m_index;
+    //!\brief Pointer to the indexed text.
+    text_t const * text = nullptr;
+
+    //!\publicsection
+
+public:
+    //!\brief The type of the indexed text.
+    using text_type = text_t;
+    //!\brief The type of the underlying character of text_type.
+    using char_type = innermost_value_type_t<text_t>;
+    //!\brief Type for representing positions in the indexed text.
+    using size_type = typename sdsl_index_type::size_type;
+
+    //!\brief The type of the (unidirectional) iterator.
+    using iterator_type = fm_index_iterator<fm_index<text_t, fm_index_traits>>;
+    friend class fm_index_iterator<fm_index<text_t, fm_index_traits>>;
+    friend class detail::fm_index_iterator_node<fm_index<text_t, fm_index_traits>>;
+
+    // static const bool is_bidirectional = false;
+    // static constexpr uint8_t dimensions = DIMENSIONS;
+
+    /*!\name Constructors and destructor
+     * \{
+     */
+    fm_index() = default;
+    fm_index(fm_index const &) = default;
+    fm_index & operator=(fm_index const &) = default;
+    fm_index(fm_index &&) = default;
+    fm_index & operator=(fm_index &&) = default;
+    ~fm_index() = default;
+    //!\}
+
+    /*!\brief Constructor that immediately constructs the index given a range.
+     *        The range cannot be an rvalue (i.e. a temporary object).
+     * \tparam text_t The type of range to construct from; must satisfy seqan3::forward_range_concept.
+     * \param[in] text The text to construct from.
+     *
+     * ### Complexity
+     *
+     * At least linear.
```

Comment:
So Omega(n) or O(n) ?

Reply:
That's not meant to be a serious documentation. Could be a TODO as well. I'll specify that later as soon as we work on the index construction. I have no idea what the running time actually is. :grimacing: